### ISSUE TITLE
Reconsider project branch structure

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#270]
Currently we're using the branch `omnicore-0.0.10` as only, and primary branch for this project.

It somehow makes sense at the moment, but after the release the name is no longer suitable.

Previously it was suggested to create a new branch, let's say `omnicore-0.0.11` afterwards, and continue the work there.

While this sounds somewhat reasonable, too, I'd like to reevaluate, whether this is our best option. The primary motivation, as far as I recall, was that using a new branch for each release cleanly seperates the code.

A similar effect is given with tags and releases though.

**I'd like to propose an alternative:**
1. there are two branches: `master` and `develop`
2. the `master` branch reflects "production-ready" code
3. the `develop` branch is the "integration branch" for development
4. new work is merged into the `develop` branch
5. once the `develop` work reaches a production-ready state, i.e. it's ready to be released, the `develop` branch is merged into `master`
6. finally a release is tagged
7. afterwards the work continues on the `develop` branch and the steps are repeated

A special note about Bitcoin Core: they maintain more than one branch in parallel, for once, to handle development and releases differently, and to preserve the ability to extend older version (i.e. backport changes). Neither of these properties seems to be suitable to us, and what I really dislike is that the branches can usually not be merged into each other (e.g. `0.11` won't merge into `0.10` cleanly).

What do you think?

Ping @msgilligan for further input.

**Comment by dexX7** [comment]
Given that the release was tagged, this topic becomes more imminent.

I'm not yet fully satisfied with the proposal above, so input would be very welcome.

Alternatively, we may continue as before, and start with a new `omnicore-0.x.x` branch.

**Comment by zathras-crypto** [comment]
Well, to be honest I'm not really a GitHub expert so I'm kind of happy to defer to you more experienced guys...

With that said, it'd be kind of nice if a user "building from source" (which I see as cloning and building our repo) got a stable release by default instead of unreleased code - which I guess implies that the master branch should contain a stable release...

Just thinking out loud...

**Comment by dexX7** [comment]
> Well, to be honest I'm not really a GitHub expert so I'm kind of happy to defer to you more experienced guys...

No worries, this is basically new to me, too.

> With that said, it'd be kind of nice if a user "building from source" (which I see as cloning and building our repo) got a stable release by default instead of unreleased code

Yes, I agree. Would be nice, to have a stable version.

@msgilligan: what's your opinion about this?

**Comment by msgilligan** [comment]
We should definitely have at least one long-running branch, maybe two. If you want you can read some articles on the well-known [GitFlow](http://nvie.com/posts/a-successful-git-branching-model/) approach and a critique, [GitFlow Considered Harmful](http://endoflineblog.com/gitflow-considered-harmful), offering a simpler approach and a [follow-up](http://endoflineblog.com/follow-up-to-gitflow-considered-harmful).

I'd lean towards the slightly simpler mechanism.

**Comment by msgilligan** [comment]
OK, I've thought about it some more (and re-read those articles).  Here's what I propose.

1. We create a new branch called `omni-master` from the current `omnicore-0.0.10` branch (`master` is reserved for upstream, right?)
1. Current development occurs on `omni-master`.
1. Releases are tagged on `omni-master` and if a bug-fix release needs to occur a branch can be made then.
1. The current stable release is tagged with `omni-stable`.

So, `omni-master` servers the purposes of both `master` and `develop` as originally proposed by Dexx (and me.) I think it is simpler and we probably don't need to maintain two branches. If we find that we do, we can always add `omni-develop` later.

For users who want to build the latest, stable release from source without needing to know the version (e.g. they have an automation script) they can checkout the `omni-stable` tag.

I think this is the "simplest possible solution that could work". So why not start simple?

**Comment by zathras-crypto** [comment]
Thanks bud.  What happens in this scenario when someone clones the repo and builds (ie without checking out a specific branch first)?

Eg Bitcoin Core cloning the repo and building without an explicit checkout will build the latest commit on the master branch.

**Comment by zathras-crypto** [comment]
Would setting the proposed 'omni-stable' branch as default do the job?

**Comment by msgilligan** [comment]
`omni-master` would be the default branch, so someone cloning without and explicit checkout would get the latest commit on that branch.

`omni-stable` would be a *tag* not a branch, so there is no way to set it to the default.

I think this is OK. Most people that go to Github are wanting to see the latest code on the default/master branch (importantly this is also going to include documentation pages.) But things like Jenkins servers and deployment scripts could use the `omni-stable` tag.

Is there really big issue with needing the default clone to be a stable version?

**Comment by zathras-crypto** [comment]
It's not critical, but currently anyone asking how to build Omni we can advise to simply use any of the myriad existing guides to building Bitcoin Core on the web and just replace the repo URL with ours.

It would be nice if that process resulted in stable code - but yep it's a "nice to have" rather than a requirement...

**Comment by msgilligan** [comment]
... and, generally, people expect if they clone a branch they'll get the latest unstable code, so we're no different from Bitcoin Core or any other Github project in that regard.

Always better to keep things simple unless there's a real clear need. @dexX7 what do you think?

**Comment by dexX7** [comment]
A couple of notes, to summarize the discussion so far, there are three points:

1) Use a seperate branch for Omni Core, and reserve one for "upstream" (e.g. `omni-master` + `master`):

Actually, I haven't thought about this, so I'm interested in learning why you think this could be beneficial. I assume you were referring to `bitcoin:master`? Unfortunally, it's probably not that easy, because our base doesn't derive from `bitcoin:master`, but currently `bitcoin:0.10`. Upgrading the base is a huge pain, because  the upstream branches `0.10`, `0.11` and `0.12` don't merge cleanly (see also: https://github.com/OmniLayer/omnicore/issues/75#issuecomment-160226070). If we could solve this somehow, it would be a huge achievement!

Other than that, I currently don't see a reason to mirror the base branch here, as cross-repo comparison (e.g. https://github.com/bitcoin/bitcoin/compare/0.10...OmniLayer:omnicore-0.0.10) works pretty well. But maybe I'm missing something?

2) If we use a dual branch approach, then one represents a stable version.

While Bitcoin Core does indeed use the `master` branch for the integration/development, I'm with @zathras-crypto on this one: if we use two branches, then the `master` branch should point to the stable branch. This seems more convenient, and is in line with the [GitFlow](http://nvie.com/posts/a-successful-git-branching-model/) approach.

3) Using two branches and tags is redundant.

This is a valid statement: if the `stable` branch contains the commits from the integration branch anyway, then we may as well just use tags to pin stable parts.

The upside is that it would allow to `git clone` without selecting a tag, to get a stable version, at the cost of additional complexity. What I find appealing is that we potentially could relax the branch protection policy for a `develop` branch: right now all pull requests must a) pass the CI tests (which is fine), but b) must be up-to-date with the base branch. Unfortunally GitHub bundles both, and this creates a huge bottleneck, where only one active PR can be handled, and all other outstanding PRs must be rebased afterwards to get merged. We could relax it for the `develop` branch, and use the strict policy for the `stable` (or `master`) branch.

When looking at the bigger picture, I think resolving https://github.com/OmniLayer/omnicore/issues/75#issuecomment-160226070 seems even more important, but I currently don't see a good solution.

**Comment by dexX7** [comment]
Oh, and a note regarding "the commit history can get messy" (which was mentioned in one of the articles):

I use the [contrib/devtools/github-merge.sh](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/contrib/devtools/github-merge.sh) script to merge pull requests as follows:

```
git clone https://github.com/OmniLayer/omnicore.git
cd omnicore/
git config githubmerge.repository OmniLayer/omnicore
git config githubmerge.branch omnicore-0.0.10
./contrib/devtools/github-merge.sh 123 # <-- pull request number
# potentially check the result, test it etc.
# enter "exit" to confirm
# enter "m" to merge
# enter "s" to sign the pull request
# enter "push" to push
```

Frequent rebasing on top of an up-to-date base branch is also useful (instead of pulling the base branch into a feature branch).

**Comment by dexX7** [comment]
Once we release 0.0.11, the current branch omnicore-0.0.10 seems very wrong.

**Comment by dexX7** [comment]
We changed the branch structure to `master` and `develop`.
